impress-org/givewp-next-gen
# Feature: store custom fields in donor and donation meta

## Description



Currently, the form builder supplies functionality for adding custom fields, defining their name, label, and if they should store in donor/donation meta.  This PR updates the front-end donation form processing to interpret custom field data and store the values in meta.

The server has also shifted architecture to be more dynamic in the way it processes & validates data.  We are now looping through the form's schema using the fields api to compare against the request data instead of hard-coding what's expected from the request.  In short, the form's schema will tell us everything we need to know about how to process a form.

## Affects



Form submissions & validation are affected under the hood. 

## Visuals



This is a custom field error being delivered from the server:

<img width="710" alt="Screen Shot 2022-10-06 at 5 18 06 PM" src="https://user-images.githubusercontent.com/10138447/194420572-dc000979-b364-4884-9dba-cf43f1a6a31e.png">


## Testing Instructions



**Test one**
Make sure the meta information is being stored correctly and in the expected meta.

- Create a Next Gen Form
- Add 2 custom fields
- Check the "Store as Donor Meta" for one of them
- Submit the form
- Check you're database for the 2 meta fields and their values in donation and donor meta

**Test two**
At the time of writing this PR, our custom text fields do not recognize the _required_ attribute.  While this will be updated in another PR, it gives us a perfect opportunity to make sure the server responds with field level errors.  

- Create a Next Gen Form
- Add a custom field that is marked as required.
- Try submit the form with that field left empty
- That field should be highlighted with an error.

## Pre-review Checklist
